packages/core/src/defer/triggering.ts

```diff
@@ -471,69 +467,78 @@ function onDeferBlockCompletion(lDetails: LDeferBlockDetails, callback: VoidFunc
- *  - on the client for blocks that were server-side rendered, to start hydration process
- */
-export function shouldActivateHydrateTrigger(lView: LView, tNode: TNode): boolean {
-  const lDetails = getLDeferBlockDetails(lView, tNode);
-  const injector = lView[INJECTOR]!;
-  // TODO(incremental-hydration): ideally, this check should only happen once and then stored on
-  // LDeferBlockDetails as a flag. This would make subsequent lookups very cheap.
-  return (
-    isIncrementalHydrationEnabled(injector) &&
-    ((typeof ngServerMode !== 'undefined' && ngServerMode) || lDetails[SSR_UNIQUE_ID] !== null)
-  );
-}
-
 // TODO(incremental-hydration): Optimize this further by moving the calculation to earlier
 // in the process. Consider a flag we can check similar to LView[FLAGS].
 
 /**
  * Determines whether regular defer block triggers should be invoked based on client state
  * and whether incremental hydration is enabled. Hydrate triggers are invoked elsewhere.
  */
-export function shouldTriggerWhenOnClient(
-  injector: Injector,
-  lDetails: LDeferBlockDetails,
-  tDetails: TDeferBlockDetails,
-): boolean {
-  if (typeof ngServerMode !== 'undefined' && ngServerMode) {
-    return false;
-  }
+export function shouldTriggerWhenOnClient(lView: LView, tNode: TNode): boolean {
+  if (typeof ngServerMode !== 'undefined' && ngServerMode) return false;
+
+  const lDetails = getLDeferBlockDetails(lView, tNode);
+  const tDetails = getTDeferBlockDetails(lView[TVIEW], tNode);
 
   const isServerRendered =
     lDetails[SSR_BLOCK_STATE] && lDetails[SSR_BLOCK_STATE] === DeferBlockState.Complete;
-  const hasHydrateTriggers = tDetails.hydrateTriggers && tDetails.hydrateTriggers.size > 0;
-  if (hasHydrateTriggers && isServerRendered && isIncrementalHydrationEnabled(injector)) {
+  const hasHydrateTriggers =
+    lDetails[SSR_UNIQUE_ID] !== null ||
+    (tDetails.hydrateTriggers !== null && tDetails.hydrateTriggers.size > 0);
+  if (hasHydrateTriggers && isServerRendered && isIncrementalHydrationEnabled(lView[INJECTOR]!)) {
     return false;
   }
   return true;
 }
 
-/**
- * Returns whether defer blocks should be triggered.
- *
- * Currently, defer blocks are not triggered on the server,
- * only placeholder content is rendered (if provided).
- */
-export function shouldTriggerDeferBlock(
-  injector: Injector,
-  tDeferBlockDetails: TDeferBlockDetails,
-): boolean {
-  const config = injector.get(DEFER_BLOCK_CONFIG, null, {optional: true});
+export function shouldTriggerDeferBlock(lView: LView, tNode: TNode) {
+  const lDetails = getLDeferBlockDetails(lView, tNode);
+  const tDetails = getTDeferBlockDetails(lView[TVIEW], tNode);
+  const incrementalHydrationEnabled = isIncrementalHydrationEnabled(lView[INJECTOR]!);
+  const hasHydrateTriggers =
+    lDetails[SSR_UNIQUE_ID] !== null ||
+    (tDetails.hydrateTriggers !== null && tDetails.hydrateTriggers.size > 0);
```

Comment:
If we rely on the `tDetails.hydrateTriggers` presence, we'll make an implicit ordering dependency between `deferHydrate*` and `defer*` (non-hydrate) instructions, i.e. hydrate ones must be invoked first. This might be a problem for `deferHydrateWhen` instruction, which would end up in the "update" block of a template function, when the other functions (e.g. `deferOnIdle`) would be in the "create" block. I'm wondering if actually checking the `SSR_UNIQUE_ID` field value presence in `LDetails` should be sufficient. That would mean that this block was triggered on the server, thus it has "hydrate" triggers. 
```suggestion
  const hasHydrateTriggers = lDetails[SSR_UNIQUE_ID] !== null;
```

Reply:
You are correct, though we do implicitly ingest hydrate triggers first in the compiler for this very reason. Perhaps we should be consistent about processing hydrate triggers first everywhere, since we do actually need to know if they are present.